File: Python/AoC2025/days/day6.py
import math

class Day6:
    def __init__(self):
        with open("../../Data/2025/day6.txt") as f:
            self.data = [line.rstrip('\n') for line in f]
            self.p1data = [line.split() for line in self.data]
            self.p2data = list(map(list, zip(*self.data)))


    def part_one(self):
        total = 0
        rows = self.p1data
    
        for i in range(len(rows[0])):
            col = [int(row[i]) for row in rows[:-1]]
            op = rows[-1][i] 
            if op == '+':
                total += sum(col)
            else:
                total += math.prod(col)
        return total

    def part_two(self):
        total = 0
        op = ' '
        nums = []
        for col in self.p2data:
            if col[-1] != ' ':
                op = col[-1]
            if all(x == ' ' for x in col):
                total += sum(nums) if op == '+' else math.prod(n for n in nums)
                nums.clear()
            else:
                nums.append(int("".join(col[:-1])))
        total += sum(nums) if op == '+' else math.prod(n for n in nums)
        return total

    def print_solution(self):
        print("Day 6:")
        print(f"Part 1: {self.part_one()}")
        print(f"Part 2: {self.part_two()}")

File: CSharp/AoC2025/Days/Day6.cs
using System.Text;

namespace AoC2025.Days;

public static class Day6
{
    public static long PartOne()
    {
        var data = GetData();
        int length = data.First().Count();
        long total = 0;

        for (int i = 0; i < length; i++)
        {
            var calculation = GetCalculation(data, i);
            total += Calculate(calculation);
        }

        return total;
    }

    public static long PartTwo()
    {
        string[] data = GetStrings();
        int length = data[0].Length;

        long total = 0;
        List<int> nums = [];
        char op = ' ';

        for (int i = 0; i < length; i++)
        {
            int result = ScanColumn(data, i, out char tempOp);
            if (tempOp != ' ')
                op = tempOp;

            if (result == -1)
            {
                total += op == '*' ? nums.Aggregate(1L, (a, b) => a * b) : nums.Sum();
                op = ' ';
                nums.Clear();
            }
            else
            {
                nums.Add(result);
            }

            if (i == length - 1)
            {
                total += op == '*' ? nums.Aggregate(1, (a, b) => a * b) : nums.Sum();
            }
        }

        return total;
    }

    // Helper methods

    private static long Calculate(IEnumerable<string> calculation)
    {
        string op = calculation.Last();
        long total = op == "*" ? 1 : 0;
        
        foreach (string num in calculation)
        {
            if (int.TryParse(num, out int n))
            {
                if (op == "*")
                    total *= n;
                else
                    total += n;
            }
        }

        return total;
    }

    private static int ScanColumn(string[] data, int index, out char op)
    {
        int length = data.Length;
        op = data[length - 1][index];
        
        if (data.All(line => line[index] == ' '))
            return -1;

        var num = new StringBuilder();
        foreach (string line in data)
        {
            if (line[index] == ' ' || line[index] == '*' || line[index] == '+')
                continue;
            num.Append(line[index]);
        }

        return int.Parse(num.ToString());
    }

    // Data methods

    private static string[] GetStrings() => File.ReadAllLines("../../Data/2025/day6.txt");

    private static IEnumerable<IEnumerable<string>> GetData() => File
        .ReadAllLines("../../Data/2025/day6.txt")
        .Select(line => line.Split(" ", StringSplitOptions.RemoveEmptyEntries));

    private static IEnumerable<string> GetCalculation(IEnumerable<IEnumerable<string>> data, int index) => data
        .Select(line => line.Skip(index).First());
}